<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

## Exercise 1

In [10]:
%%timeit
list = []

for number in range (10):
    if number%2 == 0:
        continue
    else:
        list += [number**2]

print(list)


[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 49, 81]
[1, 9, 25, 4

## Exercise 2

In [11]:
%%timeit
[number**2 for number in range(10) if number%2 == 1]

1.61 μs ± 46.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


## Exercise 3
Using list comprehension in Exercise 2 was faster, where the mean time, 1.61μs, was shorter than that in Exercise 1, which was 474μs.

## Exercise 4

In [13]:
amount = 100
years = 0

while amount > 50:
    amount = amount*0.95
    years = years + 1

print(f"The sample will take {years} years to reduce to half its original amount.")

The sample will take 14 years to reduce to half its original amount.


## Exercise 5

In [37]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]

for index, (year, CO2_level) in enumerate(co2_data[:-1]):
    #[:-1] gives the range of elements for the enumerate() function, where the last element id excluded
    diff = co2_data[index+1][1] - co2_data[index][1]
    if diff < 3:
        continue
    else:
        print(f"From {year} to {year+1}, the CO2 levels increased by {diff}ppm.")

From 2015 to 2016, the CO2 levels increased by 3.410000000000025ppm.
